pkg/internal/openshift/detector.go

```diff
@@ -0,0 +1,61 @@
+package openshift
+
+import (
+	"context"
+
+	"github.com/blang/semver/v4"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ImageVersionVariable = "OPERATOR_IMAGE_VERSION"
+
+// VersionDetector detects an OpenShift version.
+type VersionDetector struct {
+	// Client is a Kubernetes client.
+	Client client.Client
+
+	// PodNamespace is a namespace expected to contain pods with OpenShift version information.
+	PodNamespace string
+
+	// PodSelector is a label selector for listing pods with OpenShift version information.
+	PodSelector labels.Selector
+}
+
+// NewVersionDetector creates a new VersionDetector.
+func NewVersionDetector(
+	client client.Client,
+	namespace string,
+	selector labels.Selector,
+) (*VersionDetector, error) {
+	return &VersionDetector{
+		Client:       client,
+		PodSelector:  selector,
+		PodNamespace: namespace,
+	}, nil
+}
+
+// DetectVersion detects the OpenShift version and returns it and a boolean. The boolean is true if detection
+// succeeded, or false if the detector could not find a version.
+func (d *VersionDetector) DetectVersion(ctx context.Context) (semver.Version, bool) {
+	result := semver.Version{}
```

Comment:
Personally, I prefer to return a variable initialized in place instead of creating one and then relying whether someone changed it.

This way, when you return a variable when you did not find what you were looking for,

```
return semver.Version{}, false
```

becomes much more obvious than

```
return result, false
```

in the aspect of "a thing has not been found".